### ISSUE TITLE
Dozens of "No message string for " warning

### ISSUE THREAD

**Issue by NCC1701M** [roroettg/ngx-blockly#104]
I'm trying out version 14.0.0-RC1 and when I proceed as described in the tutorial I get the dozens of the `No message string for ` warning in the browsers console.

Did I miss something? Do I have to install and or link another javascript module? Maybe directly from Blockly?

![grafik](https://user-images.githubusercontent.com/7260053/204806112-ec461383-dc44-40e4-be5b-01dfa3682c43.png)

**Comment by roroettg** [comment]
Strange, does this message appear for every Block you are using? Have you changed the language of blockly? Could you show how you are initialising the blockly component?

**Comment by NCC1701M** [comment]
> Strange, does this message appear for every Block you are using? Have you changed the language of blockly? Could you show how you are initialising the blockly component?

I will provide the information next week

**Comment by NCC1701M** [comment]
> does this message appear for every Block you are using?
It already happens, directly after the application is running. There is no call of the component at this point.

But I have created a custom library for using ngx-blockly because I want to use it in different projects with a default configuration and custom blocks.

I have a Library-Module:

```typescript
@NgModule({
  declarations: [
    MyBlocklyEditorComponent
  ],
  imports: [
    CommonModule,
    FormsModule,
    MatButtonModule,
    MatMenuModule,
    NgxBlocklyModule,
    ReactiveFormsModule,
    RouterModule
  ],
  exports: [
    MatButtonModule,
    MatMenuModule,
    NgxBlocklyModule
  ],
  providers: [
  ]
})
export class MyBlocklyModule { }
```

I use the ngx-blockly component in the `MyBlocklyEditorComponent` component:

```html
<div>
  <div id="editor-container" class="shell-content grid-container">
    <ngx-blockly *ngIf="showEditor" [config]="config" [customBlocks]="customBlocks"
      (workspaceChange)="onWorkspaceChanged($event)" (workspaceCreate)="onWorkspaceCreated($event)"></ngx-blockly>
  </div>
  <div>
    <button mat-raised-button [disabled]="!changed" (click)="run()">
      Run
    </button>
  </div>
</div>
```

```typescript
const blocklyConfig: NgxBlocklyConfig = {
  generators: [
    NgxBlocklyGenerator.JAVASCRIPT
  ],
  toolbox: "",
  trashcan: true

};

@Component({
  selector: "my-blockly-editor",
  templateUrl: "./blockly-editor.component.html",
  styleUrls: ["./blockly-editor.component.scss"]
})
export class MyBlocklyEditorComponent implements OnInit, OnDestroy {
  protected readonly subscriptions: Subscription[] = [];
  public config = blocklyConfig;

  public generatorConfig: NgxBlocklyGeneratorConfig = {
    javascript: true,
  };

  public changed = false;
  public showEditor = false;
  public customBlocks: CustomBlock[];

  /**
   * Creates a new instance of this class.
   * @param translator DI for the translation service.
   * @param _route  DI for the activated route.
   * @param _ngxToolboxBuilder DI for the toolbox builder service.
   * @param _structureService DI for the structure service.
   * @param _blocklyService DI for the blockly service.
   * @param _functionService DI for the blockly function service.
   */
  constructor(
    private _blocklyService: BlocklyService
  ) {
  }

  public ngOnDestroy(): void {
    this.subscriptions.forEach(sub => {
      sub.unsubscribe();
    });
  }

  public onWorkspaceCreated(workspace: Blockly.WorkspaceSvg) {
    // Do something here
  }

  public onWorkspaceChanged(evt: Blockly.Events.AbstractEventJson) {
    if (evt instanceof Blockly.Events.FinishedLoading) {
      this.changed = false;
    } else if (evt instanceof Blockly.Events.BlockChange || evt instanceof Blockly.Events.BlockCreate || evt instanceof Blockly.Events.BlockDelete) {
      this.changed = true;
    } else if (evt instanceof Blockly.Events.BlockDrag) {
      this.changed = this.changed || !evt.isStart;
    }
  }

  protected subscribe() {
    this.subscriptions.push(
      this._blocklyService.onCustomBlockyLoaded.subscribe(
        next => {
          this.customBlocks = next;
          this.createToolbox();
        }
      )
    );
  }

  /** Creates the toolbox. */
  private createToolbox() {
    const workspace = new Blockly.WorkspaceSvg(new Blockly.Options({}));
    const toolbox = new NgxBlocklyToolbox(workspace);

    toolbox.nodes = [
      new MyBlockyCategory(this.customBlocks),
      new Separator(),
      LOGIC_CATEGORY,
      LOOP_CATEGORY,
      MATH_CATEGORY,
      TEXT_CATEGORY,
      LISTS_CATEGORY,
      VARIABLES_CATEGORY
    ];
    this.config.toolbox = toolbox.toXML();
    this.showEditor = true;
  }
}
```

The configuration in the `angular.json` for the library looks like this:

```json
{
  "$schema": "./node_modules/@angular/cli/lib/config/schema.json",
  "version": 1,
  "newProjectRoot": "projects",
  "projects": {
    "my-blockly-library": {
      "projectType": "library",
      "root": "projects/my-blockly-library",
      "sourceRoot": "projects/my-blockly-library/src",
      "prefix": "lib",
      "architect": {
        "build": {
          "builder": "@angular-devkit/build-angular:ng-packagr",
          "options": {
            "tsConfig": "projects/my-blockly-library/tsconfig.lib.json",
            "project": "projects/my-blockly-library/ng-package.json"
          },
          "configurations": {
            "production": {
              "tsConfig": "projects/my-blockly-library/tsconfig.lib.prod.json"
            },
            "development": {}
          },
          "defaultConfiguration": "production"
        },
        "lint": {
          "builder": "@angular-devkit/build-angular:tslint",
          "options": {
            "tsConfig": [
              "projects/my-blockly-library/tsconfig.lib.json"
            ],
            "exclude": [
              "**/node_modules/**"
            ]
          }
        }
      }
    }
  }
}
```

I implement the library in an extra module in the application:

```typescript
@NgModule({
  imports: [
    CommonModule,
    FormsModule,
    ReactiveFormsModule,
    HttpClientModule,
    MaterialModule,
    AdminRoutingModule,
    MyBlocklyModule,
    NgxBlocklyModule
  ],
  declarations: [
    RunBlocklyComponent
  ]
})
export class AdminModule { }
```

The editor will be used in the `RunBlocklyComponent` using `<my-blockly-editor></my-blockly-editor>`.

For the application I've installed the package `blockly` in version `^6.20210701.0` as well as `ngx-blockly` in version `^14.0.0-RC1` and of course my library.

I've added the following code to the applications `angular.json`:
```json
"architect": {
  "build": {
    "builder": "@angular-devkit/build-angular:browser",
    "options": {
      "outputPath": "dist",
      "index": "src/index.html",
      "main": "src/main.ts",
      "polyfills": "src/polyfills.ts",
      "tsConfig": "tsconfig.app.json",
      "assets": [
        "src/favicon.ico",
        "src/assets",
      ],
      "styles": [
        "src/styles.scss",
      ],
      "scripts": [
      ],
      "allowedCommonJsDependencies": [
        "angular2-uuid",
        "blockly/blocks",
        "blockly/core",
        "blockly/python",
        "blockly/php",
        "blockly/lua",
        "blockly/javascript",
        "blockly/dart",
        "blockly/msg/en",
        "blockly/msg/de"
      ],
      "extractLicenses": true
    },
    // ...
  }
}
```

The code above is not complete. There are more components in my library and the edit component has more functionality. But it's basically my implementation. It worked well with version `^13.0.1` of `ngx-blockly` and I didn't change anything except the version.

**Comment by roroettg** [comment]
Hi,

you dont need to have blockly as a dependency since it is shipped with ngx-blockly. 

ngx-blockly tries to import as less as possible from blockly itself to ensure not importing stuff a consumer wont need.

That is why ngx-blockly only imports 'blockly/core', which does not contain blocklys default blocks. You will need to import 'blockly/blocks' somewhere in your code to have them. I would recommend to import them as early as possible to avoid conflicts.

Please let me know, if that solved the issue.


FYI: https://www.npmjs.com/package/blockly

**Comment by NCC1701M** [comment]
Ok, I've already did the `import "blockly/blocks"` but it seems it was to late in the code. I moved it to the module of my library and now the `No message string for` warnings are gone.

But now I get another error when dragging or dropping a block (also the default blocks) to the workspace:

```
core.mjs:7635 
ERROR TypeError: Cannot read properties of undefined (reading 'workspaceToCode')
    at NgxBlocklyComponent.workspaceToCode (ngx-blockly.mjs:158:86)
    at NgxBlocklyComponent._onWorkspaceChange (ngx-blockly.mjs:284:22)
    at WorkspaceSvg$$module$build$src$core$workspace_svg.fireChangeListener (blockly_compressed.js:1093:309)
    at fireNow$$module$build$src$core$events$utils (blockly_compressed.js:68:298)
    at timer (zone.js:3155:47)
    at _ZoneDelegate.invokeTask (zone.js:443:35)
    at Object.onInvokeTask (core.mjs:26218:33)
    at _ZoneDelegate.invokeTask (zone.js:442:64)
    at Zone.runTask (zone.js:214:51)
    at ZoneTask.invokeTask (zone.js:525:38)
```
The error seems to be thrown here:
```javascript
this.javascriptCode.emit(Blockly[NgxBlocklyGenerator.JAVASCRIPT].workspaceToCode(Blockly.Workspace.getById(workspaceId)));
```

I tried to fix this by adding `import "blockly/javascript";` directly after the blocks import but it doesn't help. Did I miss something or should I open an issue for this?

**Comment by NCC1701M** [comment]
> But now I get another error when dragging or dropping a block (also the default blocks) to the workspace:
> 
> ```
> core.mjs:7635 
> ERROR TypeError: Cannot read properties of undefined (reading 'workspaceToCode')
>     at NgxBlocklyComponent.workspaceToCode (ngx-blockly.mjs:158:86)
>     at NgxBlocklyComponent._onWorkspaceChange (ngx-blockly.mjs:284:22)
>     at WorkspaceSvg$$module$build$src$core$workspace_svg.fireChangeListener (blockly_compressed.js:1093:309)
>     at fireNow$$module$build$src$core$events$utils (blockly_compressed.js:68:298)
>     at timer (zone.js:3155:47)
>     at _ZoneDelegate.invokeTask (zone.js:443:35)
>     at Object.onInvokeTask (core.mjs:26218:33)
>     at _ZoneDelegate.invokeTask (zone.js:442:64)
>     at Zone.runTask (zone.js:214:51)
>     at ZoneTask.invokeTask (zone.js:525:38)
> ```
> The error seems to be thrown here:
> ```javascript
> this.javascriptCode.emit(Blockly[NgxBlocklyGenerator.JAVASCRIPT].workspaceToCode(Blockly.Workspace.getById(workspaceId)));
> ```
> 
> I tried to fix this by adding `import "blockly/javascript";` directly after the blocks import but it doesn't help. Did I miss something or should I open an issue for this?

@roroettg Can you give me a response in that question?

**Comment by roroettg** [comment]
Hi @NCC1701M,

sorry for letting you wait on this. I was really busy lately.

This probably is an issue in the RC you are using. I have just uploaded version 14.0.0 to npm. Please check if the issue still is present.

**Comment by NCC1701M** [comment]
Hi @roroettg 

version 14.0.0 seams to fix the problems with the default blocks but now I unfortunately get error messages with my custom blocks:

```
ERROR Error: Language "JavaScript" does not know how to generate code for block type "current_date_block".
    at $.CodeGenerator$$module$build$src$core$generator.blockToCode (blockly_compressed.js:1330:165)
    at $.CodeGenerator$$module$build$src$core$generator.workspaceToCode (blockly_compressed.js:1328:236)
    at NgxBlocklyComponent.workspaceToCode (ngx-blockly.mjs:158:46)
    at NgxBlocklyComponent._onWorkspaceChange (ngx-blockly.mjs:284:22)
    at WorkspaceSvg$$module$build$src$core$workspace_svg.fireChangeListener (blockly_compressed.js:1093:309)
    at fireNow$$module$build$src$core$events$utils (blockly_compressed.js:68:298)
    at timer (zone.js:3155:47)
    at _ZoneDelegate.invokeTask (zone.js:443:35)
    at Object.onInvokeTask (core.mjs:26218:33)
    at _ZoneDelegate.invokeTask (zone.js:442:64)
```

The code for my custom `current_date_block` is:

```typescript
import { Blockly, BlockMutator, CustomBlock } from "ngx-blockly";

export class CurrentDateBlock extends CustomBlock {

	constructor(
		type: string,
		blockMutator: BlockMutator = null,
		...args: any[]
	) {
		super(type, blockMutator, ...args);
		this.class = CurrentDateBlock;
	}

	public defineBlock() {
		this.block.appendDummyInput()
			.appendField("Current Date");
		this.block.setOutput(true, "Date");
		this.block.setColour("#ac7439");
	}

	public toJavaScriptCode(_: Blockly.Block): string | any[] {
		return "new Date()";
	}
}
```

**Comment by roroettg** [comment]
Hi @NCC1701M,

don't know, how I missed that yesterday, but the release had another issue and I have published a new version just now, which should resolve that issue. 

Sorry for the inconveniences this might have caused.

**Comment by MBAatEJOT** [comment]
Now I get an error when compiling the library where I use ngx-blockly

```
✖ Compiling with Angular sources in Ivy partial compilation mode.
node_modules/ngx-blockly/lib/ngx-blockly/ngx-blockly.component.d.ts:80:18 - error TS2707: Generic type 'ɵɵComponentDeclaration' requires between 7 and 8 type arguments.

80     static ɵcmp: i0.ɵɵComponentDeclaration<NgxBlocklyComponent, "ngx-blockly", never, { "config": "config"; "customBlocks": "customBlocks"; "readOnly": "readOnly"; }, { "workspaceCreate": "workspaceCreate"; "workspaceChange": "workspaceChange"; "toolboxChange": "toolboxChange"; "dartCode": "dartCode"; "javascriptCode": "javascriptCode"; "luaCode": "luaCode"; "phpCode": "phpCode"; "pythonCode": "pythonCode"; "xmlCode": "xmlCode"; }, never, never, false, never>;
```

I don't know if this is relevant but in your package.json you target Angular 14 as expected because I use Angular 14 but your package-lock.json has been updated to angular 15

![image](https://user-images.githubusercontent.com/70940842/214773046-af5c32e3-9095-474d-a5b3-1f8a1439221d.png)

**Comment by roroettg** [comment]
Hi @MBAatEJOT,

thank you for your feedback. You are completely right the lib was built with the wrong angular version. I still need to figure out a way to supply multiple versions of the lib for different angular versions.

With 14.0.2 everything should hopefully be fixed.

**Comment by MBAatEJOT** [comment]
Hi @roroettg 

thanks for the response. Now it's working.

**Comment by NCC1701M** [comment]
It's working for me, too.
